### ISSUE TITLE
[Question] How to check for Brand Type

### ISSUE THREAD

**Issue by ortegamarcel** [kourge/ts-brand#9]
If a parameter can be one of two different brands. How can I check which one was passed?

Example:

```typescript
export type Id = Brand<string, 'id'>;
export type AlternativeId = Brand<string, 'alt_id'>;

function loadById(id: Id | AlternativeId) {
  if (id instanceof Id) {
    // Do something
  } else if (id instanceof AlternativeId) {
    // Do something
  } else {
    throw new Error('unsupported id type');
  }
}
```

**Comment by kourge** [comment]
Unfortunately, this kind of check is not possible with branded types alone, because they will both be just strings at runtime. To make it possible to check, you'd need to pass around what type you know them to be. This can be accomplished with a [discriminated union](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#discriminated-unions), also known as [tagged unions](https://mariusschulz.com/blog/tagged-union-types-in-typescript):

```ts
export type Id = Brand<string, 'id'>;
export type AlternativeId = Brand<string, 'alt_id'>;
export type AnyId = (
  | { type: 'id'; value: Id }
  | { type: 'alt_id'; value: AlternativeId }
);

function loadById(id: AnyId) {
  if (id.type === 'id') {
    // Do something to `id.value`
  } else if (id.type === 'alt_id') {
    // Do something to `id.value`
  } else {
    throw new Error('unsupported id type');
  }
}
```

If you find yourself needing, at runtime, to tell apart different branded types, then branded types are almost certainly not the right solution for solving your problem. The above example, for instance, doesn't need branded types to work at all. I could replace them with plain strings and it would work just as well. Like this:

```ts
export type AnyId =
  | { type: 'id', value: string }
  | { type: 'alt_id': value: string }
;

function loadById(id: AnyId) {
  if (id.type === 'id') {
    // Do something to `id.value`
  } else if (id.type === 'alt_id') {
    // Do something to `id.value`
  } else {
    throw new Error('unsupported id type');
  }
}
```

In other words, if you have a bunch of strings with different flavors and you need to treat them differently depending on the flavor, you need discriminated unions. But if you have a bunch of strings with different flavors, and you don't want to ever accidentally treat one flavor as another, or worse, one flavor as a generic string, then branded types can solve this problem.
